irinterp: add Tarjan SCC algorithm for reachability

topolarity opened the pull request:
This PR optimizes `kill_edge!` for IR interp.

The basic algorithm flow is:
```
   1. Check whether `target` of dead edge is unreachable, which is true iff:
       - Reducible CFG node: there are no live incoming forward edges (predecessors)
       - Irreducible CFG node: Tarjan's SCC algorithm reports no live incoming forward edges to the SCC
   2. If `target` is dead, repeat (1) for all of its outgoing edges
```

This maintains reachability information very efficiently, especially for reducible CFG's which are overwhelmingly common.

As an added bonus, `CFGReachability` can also be consulted to check which BasicBlocks are part of an irreducible loop.

topolarity committed:
irinterp: Add Tarjan SCC algorithm for reachability

This optimizes `kill_edge!` for IR interp. The basic algorithm flow is:

   1. Check whether `target` of dead edge is unreachable, i.e. iff:

       - Reducible CFG node: there is no live incoming forward edge

       - Irreducible CFG node: Tarjan's SCC algorithm reports no live
         incoming forward edges to the SCC

   2. If `target` is dead, repeat (1) for all of its outgoing edges

This maintains reachability information very efficiently, especially
for reducible CFG's which are overwhelmingly common. As an added bonus
CFGReachability can also be consulted to check which BasicBlocks are
part of an irreducible loop.

topolarity commented:
This is 95% ready for review, but there's still a bit of clean-up and a test or two that I want to add before merging.

Keno commented on base/compiler/ssair/tarjan.jl:
```diff
@@ -0,0 +1,299 @@
+struct CFGReachability
+    irreducible::BitVector # BBNumber -> Bool
+    scc::Vector{Int}       # BBNumber -> SCCNumber
+    domtree::DomTree
```
What keeps this domtree in sync during edge removal?

topolarity replied:
It's intentionally kept out of sync: maintaining the dominance tree is O(N^2) at least while this is O(N) (for reducible CFGs) so there's a performance motivation, but there's a deeper reason w.r.t. SSA legality: I am pretty sure that if you want to use the dominance tree to locate phi nodes, you actually want to keep it out of sync.

My reasoning is that after edge deletion if you update the domtree, you'll remove dominance frontiers where you still have (single-element) ϕ nodes. Eliminating those ϕ nodes requires both updating the domtree and re-ordering the basic blocks and all instructions to match, which is quite expensive.

In order words, updating the DomTree seems like it requires re-legalizing the SSA since the latter is defined in terms of the former.

topolarity replied:
Keeping it out of sync like this means that we keep these single-element PhiNodes around and potentially don't notice some irreducible control flow that actually became reducible.

It also means that any passes reasoning about dominance directly (rather than following use-def flow) will take a hit to precision, but I'm not sure if we do that anywhere

Keno replied:
I think it would make me happier if rather than keeping the domtree, we just kept a list of which edges we consider the backedges. It'd probably be faster too, since dominates is not quite O(1).

topolarity replied:
I think that sidesteps the issue, since external abstract interpreters might use the domtree downstream (as we do now). I'd prefer to align on what domtree we need to provide downstream, and make this work for that.

On the other hand, I'm still not convinced the updated domtree would be correct for many downstream uses (unless SSA is re-legalized). Consider for example:
- Inserting a new def-like: If the domtree reflects the true dominance rather than the dominance at SSA construction, the `iterated_dominance_frontier` will fail to report blocks where a single-element phi is required due to stmt ordering.
- Walking to all uses of stmt that flows through PhiNodes e.g. https://github.com/JuliaLang/julia/blob/47d31acd1494dfed4f8f361d67fe1fb8c3888656/base/compiler/optimize.jl#L903-L908 If the domtree is updated after edge deletions, this will fail to mark inconsistent single-element PhiNodes that propagate results from the GotoIfNot.

topolarity replied:
Also it's generally sound to under-approximate dominance relations like this - since it's equivalent to assuming the presence of control-flow edges that don't get exercised at runtime.

I do agree that _insertions_ of new control flow edges (of the same kind that require SSA re-legalization anyway) do need some kind of update to the domtree though

topolarity replied:
I was wondering why we haven't run into this already, and it turns out this is an existing bug: https://github.com/JuliaLang/julia/issues/53013

vtjnash commented on base/compiler/ssair/tarjan.jl:
```diff
@@ -0,0 +1,299 @@
+    live = _bb_externally_reachable(reach, cfg, root)
+
+    # the original algorithm has a separate stack and worklist (unrelated to `reach._worklist`)
+    # here we use a single combined stack for improved memory/cache efficiency
```
This doesn't sound like it should be possible, but I guess the argument here is that there is a second implied stack allocated as the list of `parent` links, so it is there, but represented implicitly?

topolarity replied:
Yeah, that's right - the DFS worklist is a strict subset of the stack so we maintain it implicitly with parent links here

topolarity commented:
> I'm quite suspicious of the use of the domtree in this. Domtree by itself already maintains reachability, so if the domtree is used internally, it needs to be maintained, so this will be no faster. What I would have expected here is an implementation of decremental scc maintenance (see https://theory.stanford.edu/~virgi/cs267/papers/lacki-dec-scc.pdf for the description of the algorithm - there's also more recent algorithms that are asymptotically better but way more complicated https://arxiv.org/abs/1901.03615) and no domtree.

I'm not 100% sure it's the same as the paper you link, but this is a decremental SCC maintenance algorithm (the key optimization is that when removing an edge it only traverses nodes in the SCC of the target block).

The novel trick is that if you know that a dominates b, then we can safely ignore back-edges `b -> a` when considering reachability  in that graph (because the entry block reaches `a` through that edge only if it can reach `a` through some other edge).

Since removing edges only adds dominance relations, there's no requirement to keep the domtree up-to-date

Keno commented on base/compiler/ssair/tarjan.jl:
```diff
@@ -0,0 +1,299 @@
+
+Outputs:
+  - `reach.scc`: strongly-connected components, ignoring backedges to (natural) loops
+  - `reach.irreducible`: true iff a BasicBlock is part of a (non-trivial) SCC / irreducible loop
```
Isn't this all loops, not just the irreducible ones?

Keno commented:
Alright, I've finally understood what you're trying to do here. What you're maintaining is basically the reducible condensation of the CFG.

topolarity commented:
Yeah, that sounds right.

Just to put it in other words, this is computing the condensation of the CFG with reducible back-edges removed. Probably could have said that more clearly, sorry 😅

Keno reviewed (approved):
I think this is sensible. At some point, I'd like some more docs and explanations (probably in the docstring of CFGReachability). I'd also like to see a refactor of the tarjan code to split out the filter explicitly (or maybe we should define a graph abstraction and then have a filtered version). But since this solves immediate performance issues, I think this is fine to go in.

topolarity committed:
Add `filter` parameter to `tarjan!`

This still isn't an entirely generic version of the algorithm, but it's
certainly a step in the right direction.

topolarity commented:
I'm going to merge this for now, and we can circle back on the domtree-related questions when we resolve #53013 and #53011

topolarity commented:
Also just to record a few benchmarks:
```julia
     CFG size |    old     |   new     | speedup
     6 blocks | 2.922e-06  | 5.79e-07  | (5.04663x faster)
    25 blocks | 5.174e-06  | 3.48e-07  | (14.8678x faster)
   134 blocks | 3.7766e-05 | 6.58e-07  | (57.3951x faster)
 49500 blocks | 0.00761317 | 2.852e-06 | (2669.42x faster)
```

These are the average edge-deletion times for a few different IR's - mostly from the sysimage, except for the last one which is a downstream application with very large functions.

The smaller cases are overwhelmingly more common, so this won't make a significant difference for most normal functions.
